pytools/graph.py

```diff
@@ -236,6 +237,75 @@ def compute_topological_order(graph):
 
     return list(reversed(reverse_order))
 
+
+def compute_topological_order_v2(graph, key=None):
+    """Compute a toplogical order of nodes in a directed graph.
+
+    :arg graph: A :class:`collections.abc.Mapping` representing a directed
+        graph. The dictionary contains one key representing each node in the
+        graph, and this key maps to a :class:`collections.abc.Iterable` of its
+        successor nodes.
+
+    :arg key: A custom key function may be supplied to determine the order in
+        break-even cases. Expects a function of one argument that is used to
+        extract a comparison key from each node of the *graph*.
+
+    :returns: A :class:`list` representing a valid topological ordering of the
+        nodes in the directed graph.
+
+    .. note::
+
+        * Has stricter requirements on the provided *graph* than
+          :func:`pytools.graph.compute_topological_order`: it requires the
+          keys of the mapping *graph* to be hashable and inequality comparable.
+        * Implements Kahn's algorithm.
+    """
+    if key is None:
+        def key(x):
+            # all nodes have the same keys when not provided
+            return 0
+
+    from heapq import heapify, heappop, heappush
+
+    order = []
+
+    # {{{ compute nodes_to_num_predecessors
+
+    nodes_to_num_predecessors = {node: 0 for node in graph}
+
+    for node in graph:
+        for child in graph[node]:
+            nodes_to_num_predecessors[child] += 1
+
+    # }}}
+
+    # heap: list of pairs of the form (key(n), n) where 'n' is a node in
+    # 'graph' with no predecessor. Nodes with no predecessors are the
+    # schedulable candidates.
+    heap = [(key(n), n) for n, num_preds in nodes_to_num_predecessors.items()
+            if num_preds == 0]
+    heapify(heap)
```

Comment:
I suspect this might fail if `n1 < n2` is not defined/errors for two nodes. (Since tuples compare lexicographically.) This is easily fixed with a custom class that overrides comparison.